nn/optimizer.py

```diff
@@ -0,0 +1,222 @@
+import numpy as np
+
+"""
+1. The Optimizer class is an abstract base class that defines the interface for all optimizers.
+2. It has two abstract methods update_weights and update_bias that are implemented by its subclasses.
+3. The SGD class is a subclass of Optimizer that implements the stochastic gradient descent optimizer.
+4. Its constructor takes a learning_rate parameter and calls the constructor of its superclass Optimizer.
+5. The update_weights method of SGD updates the weights of a layer by subtracting the product of the learning rate and the gradient of the weights.
+6. The update_bias method of SGD updates the bias of a layer by subtracting the product of the learning rate and the gradient of the bias.
+7. The Adam optimizer is an extension of the stochastic gradient descent optimizer that uses adaptive learning rates for each weight parameter. It also includes momentum terms and a bias correction to the adaptive learning rate.
+8. The RMSprop optimizer uses a moving average of the squared gradient to adapt the learning rate for each weight parameter.
+"""
+
+class Optimizer:
+
+    """Base optimizer class for neural network optimization.
+    
+    Parameters:
+    learning_rate (float): The learning rate to be used for optimization.
+    """
+
+    def __init__(self, learning_rate: float):
+        self.learning_rate = learning_rate
+        
+    def update_weights(self, layer, grad_weights):
+        """Update the weights of the given layer using the gradient of weights.
+        
+        Parameters:
+        layer: A layer object whose weights need to be updated.
+        grad_weights (np.ndarray): The gradient of weights computed during backpropagation.
+        
+        Raises:
+        TypeError: If grad_weights is not a NumPy array.
+        """
+        if not isinstance(grad_weights, np.ndarray):
+            raise TypeError("grad_weights should be a NumPy array")
+        weights = layer.weights

```

Comment:
For updating weights and biases in the Optimiser , you will have to do:

```python
layer.weights -= self.learning_rate * grad_weights
```

instead, and the getter and setter added via properties will do the rest

Reply:
I have fix the weights and bias assignment issue, as well as added some more tests for Adam, RMSprop and for the loss and activation functions.